### PR TITLE
Writing Token Files Update Individually with Go Routine

### PR DESCRIPTION
**Pull request by y-myajima**
# Description
Updated the internal logic of the function that outputs Access Tokens and Role Tokens to files.
The process of updating the memory cache for each token will also include file output processing. This will result in the following:
- File output and retry processing will be performed for each token.
- File output processing will be executed in parallel.

## Assignees
- [x] `Assignees` is set

## Type of changes
- [x] Apply one or more `labels` of the following that fits:
  - `bug`: Bug fix
  - `dependencies`: Dependency upgrades
  - `documentation`: Documentation changes
  - `enhancement`: New Feature
  - `good first issue`: First contribution
  - `logging`: Log changes
  - `refactor`: Refactoring (no functional changes, no api changes)

### Flags
```
- [ ] Breaks backward compatibility
- [ ] Requires a documentation update
- [ ] Has untestable code
```
---

## Checklist
```
- [ ] Followed the guidelines in the CONTRIBUTING document
- [ ] Added prefix `[skip ci]`/`[ci skip]`/`[no ci]`/`[skip actions]`/`[actions skip]` in the PR title if necessary
- [ ] Tested and linted the code
- [ ] Commented the code
- [ ] Made corresponding changes to the documentation
```

## Checklist for maintainer
```
- [ ] Use `Squash and merge`
- [ ] Double-confirm the merge message has prefix `[skip ci]`/`[ci skip]`/`[no ci]`/`[skip actions]`/`[actions skip]`
- [ ] Delete the branch after merge
```

### CODE REVIEW

File: pkg/token/service.go [hunk 1]
```diff
@@ -423,72 +438,105 @@ func (d *tokenService) updateToken(key CacheKey, tt mode) error {
...
-		return err
 	}
 
+	log.Infof("[New %s Token] Domain: %s, Role: %s", tokenType, domain, role)
+	log.Debugf("Saving %s Token[%d bytes] at %s", tokenType, len(rawToken), outPath)
+	if err := w.AddBytes(outPath, 0644, []byte(rawToken)); err != nil {
+		return fmt.Errorf("unable to save %s Token: %w", tokenType, err)
+	}
 	return w.Save()
 }
```
Comment (mlajkim) [inline pkg/token/service.go:541]: Again, try to avoid refactoring in an implementation PR.

You can just write a simple TODO saying,
// TODO: Refactor writing file function as it is the same for RT/AT

And create a new refactor PR based on it. (or an issue)

File: pkg/token/service.go [hunk 2]
```diff
@@ -41,9 +41,11 @@ type tokenService struct {
...
 	shutdownWg   sync.WaitGroup
 
-	group            singleflight.Group
-	accessTokenCache TokenCache
-	roleTokenCache   TokenCache
+	group               singleflight.Group
+	accessTokenCache    TokenCache
+	roleTokenCache      TokenCache
+	atTargetDomainRoles []CacheKey
+	rtTargetDomainRoles []CacheKey
```
Comment (mlajkim) [inline pkg/token/service.go:48]: Write comments what these are, and how they are different from `accessTokenCache` & `roleTokenCache`
I believe these are for the file inputs, but I dont see any comments about the differences unless I read codes.

File: pkg/token/service.go [hunk 3]
```diff
@@ -423,72 +438,105 @@ func (d *tokenService) updateToken(key CacheKey, tt mode) error {
...
-	b.Multiplier = 2
-	b.MaxElapsedTime = maxElapsedTime
-
-	return backoff.RetryNotify(ts.writeFiles, backoff.WithContext(b, ctx), func(err error, backoffDelay time.Duration) {
-		log.Errorf("Failed to write token files: %s. Retrying in %s", err.Error(), backoffDelay)
-	})
-}
-
-func (d *tokenService) writeFiles() error {
+func (d *tokenService) writeFiles(ctx context.Context, maxElapsedTime time.Duration) []error {
```
Comment (mlajkim) [inline pkg/token/service.go:441]: I understand you have renamed the function name `writeFilesWithRetry` to `writeFiles` by making retries default behavior for the writing files.

But again this is not a good thing to refactor in a PR for a new implementation of `atTargetDomainRoles` &  `rtTargetDomainRoles`
You can write a simple comment TODOs saying
Refactoring must be done in a different PR.

File: pkg/token/service.go (hunk 2 above)
Comment (mlajkim) [reply pkg/token/service.go:48]: prolly better if variable has something related to writing as files, so that we know these variables for writing files.

File: pkg/token/service.go [hunk 4]
```diff
@@ -78,13 +80,24 @@ func New(ctx context.Context, idCfg *config.IdentityConfig) (daemon.Daemon, erro
 	tokenExpiryInSecond := int(idCfg.TokenExpiry.Seconds())
 	accessTokenCache := NewLockedTokenCache("accesstoken", idCfg.Namespace, idCfg.PodName)
 	roleTokenCache := NewLockedTokenCache("roletoken", idCfg.Namespace, idCfg.PodName)
+	var atTargetDomainRoles, rtTargetDomainRoles []CacheKey
+	if tt&mACCESS_TOKEN != 0 {
```
Comment (mlajkim) [inline pkg/token/service.go:84]: Write a TODO for `tt&mACCESS_TOKEN != 0`
It would be cool if we can write it like:
```
if  tt.isAccessTokenEnabled() { ... }
```

**Review by mlajkim — COMMENTED** [review]
review

File: pkg/token/service.go [hunk 5]
```diff
@@ -423,72 +438,105 @@ func (d *tokenService) updateToken(key CacheKey, tt mode) error {
...
+	operation := func() error {
+		return d.writeFile(token, outPath, tt)
+	}
+	notifyOnErr := func(err error, backoffDelay time.Duration) {
+		log.Errorf("Failed to output the token to a file: %s. Retrying in %s", err.Error(), backoffDelay)
+	}
+	return backoff.RetryNotify(operation, backoff.WithContext(b, ctx), notifyOnErr)
+}
+
+func (d *tokenService) writeFile(token Token, outPath string, tt mode) error {
```
Comment (mlajkim) [inline pkg/token/service.go:515]: Write a comment for what this does
i.e)
```
// writeFile outputs given token (AT or RT) as file
```

File: pkg/token/service.go [hunk 6]
```diff
@@ -423,72 +438,105 @@ func (d *tokenService) updateToken(key CacheKey, tt mode) error {
...
+	// collect errors
+	close(echan)
+	errs := make([]error, 0, len(d.atTargetDomainRoles)+len(d.rtTargetDomainRoles))
+	for err := range echan {
+		errs = append(errs, err)
+	}
+	return errs
+}
+
+func (d *tokenService) writeFileWithRetry(ctx context.Context, maxElapsedTime time.Duration, token Token, outPath string, tt mode) error {
```
Comment (mlajkim) [inline pkg/token/service.go:499]: Write a comment for what this does
i.e)
```
// writeFileWithRetry attempts multiple times to write a file for the given token (AT or RT) by utilizing writeFile()
```

File: pkg/token/service.go (hunk 3 above)
Comment (mlajkim) [inline pkg/token/service.go:441]: Can you write a comment for this function what it does?
- Runs go routine for each target output for token, so that each token can be synced in different time manner

File: pkg/token/service.go (hunk 2 above)
Comment (y-myajima) [reply pkg/token/service.go]: fixed in https://github.com/AthenZ/k8s-athenz-sia/pull/131/commits/1e2206eb193c0553c182a0d487bf41e1f03ff7e5

File: pkg/token/service.go (hunk 4 above)
Comment (y-myajima) [reply pkg/token/service.go:86]: fixed in https://github.com/AthenZ/k8s-athenz-sia/pull/131/commits/b6d54d1aebe18e5f07b4c2f24615e520d46d0169

File: pkg/token/service.go (hunk 5 above)
Comment (y-myajima) [reply pkg/token/service.go:525]: fixed in https://github.com/AthenZ/k8s-athenz-sia/pull/131/commits/6822928c8eef4cfb5ba60157b5141e6bcf66cfe1

File: pkg/token/service.go (hunk 6 above)
Comment (y-myajima) [reply pkg/token/service.go:508]: fixed in https://github.com/AthenZ/k8s-athenz-sia/pull/131/commits/6822928c8eef4cfb5ba60157b5141e6bcf66cfe1

File: pkg/token/service.go (hunk 3 above)
Comment (y-myajima) [reply pkg/token/service.go:449]: fixed in https://github.com/AthenZ/k8s-athenz-sia/pull/131/commits/6822928c8eef4cfb5ba60157b5141e6bcf66cfe1

Comment (mlajkim) [reply pkg/token/service.go:449]: @y-myajima Usually in Go lang, you write comments with function name + verb 

Here is a sample:
```go
// RemoveAll removes path and any children it contains.
// It removes everything it can but returns the first error
// it encounters. If the path does not exist, RemoveAll
// returns nil (no error).
// If there is an error, it will be of type *PathError.
func RemoveAll(path string) error {
	return removeAll(path)
}
```

So for this function you should write like:
```
writeFiles outputs every access token and role token cached in the memory.
...

Comment (mlajkim) [reply pkg/token/service.go:449]: sample:
```
/opt/homebrew/Cellar/go/1.22.2/libexec/src/os/path.go
```

File: pkg/token/service.go [hunk 7]
```diff
@@ -423,72 +443,110 @@ func (d *tokenService) updateToken(key CacheKey, tt mode) error {
...
+	return errs
+}
+
+// writeFileWithRetry attempts multiple times to write a file for the given token (AT or RT) by utilizing writeFile()
+func (d *tokenService) writeFileWithRetry(ctx context.Context, maxElapsedTime time.Duration, token Token, outPath string, tt mode) error {
+	// backoff config with first retry delay of 5s, and then 10s, 20s, ...
+	b := backoff.NewExponentialBackOff()
+	b.InitialInterval = 5 * time.Second
+	b.Multiplier = 2
+	b.MaxElapsedTime = maxElapsedTime
```
Comment (mlajkim) [inline pkg/token/service.go:513]: Can you write a function like this:  https://github.com/AthenZ/k8s-athenz-sia/blob/main/pkg/certificate/service.go#L375-L383

and apply for all backoff used within the file?

File: pkg/token/service.go [hunk 8]
```diff
@@ -133,7 +153,7 @@ func New(ctx context.Context, idCfg *config.IdentityConfig) (daemon.Daemon, erro
 		for _, err := range errs {
 			log.Errorf("Failed to refresh tokens after multiple retries: %s", err.Error())
 		}
-		if err := ts.writeFilesWithRetry(ctx, config.DEFAULT_MAX_ELAPSED_TIME_ON_INIT); err != nil {
+		for _, err := range ts.writeFiles(ctx, config.DEFAULT_MAX_ELAPSED_TIME_ON_INIT) {
```
Comment (mlajkim) [inline pkg/token/service.go:156]: Log only if it is err?

Comment (mlajkim) [reply pkg/token/service.go:156]: `err != nil`

File: pkg/token/service.go [hunk 9]
```diff
@@ -223,7 +243,7 @@ func (ts *tokenService) Start(ctx context.Context) error {
 						log.Errorf("Failed to refresh tokens after multiple retries: %s", err.Error())
 					}
 					// backoff retry until TOKEN_REFRESH_INTERVAL / 4 OR context is done
-					if err := ts.writeFilesWithRetry(ctx, ts.tokenRefresh/4); err != nil {
+					for _, err := range ts.writeFiles(ctx, ts.tokenRefresh/4) {
```
Comment (mlajkim) [inline pkg/token/service.go:246]: Log only if it is `err != nil`

**Review by mlajkim — COMMENTED** [review]
review

File: pkg/token/service.go (hunk 8 above)
Comment (mlajkim) [reply pkg/token/service.go:156]: my wrongful understanding

File: pkg/token/service.go (hunk 9 above)
Comment (mlajkim) [reply pkg/token/service.go:246]: my wrongful understanding

File: pkg/token/service.go [hunk 10]
```diff
@@ -423,72 +437,104 @@ func (d *tokenService) updateToken(key CacheKey, tt mode) error {
...
+	if rawToken == "" {
+		// skip placeholder token added during daemon creation
 		return nil
-	})
-	if err != nil {
-		return err
 	}
 
+	log.Infof("[New %s Token] Domain: %s, Role: %s", tokenType, domain, role)
+	log.Debugf("Saving %s Token[%d bytes] at %s", tokenType, len(rawToken), outPath)
```
Comment (mlajkim) [inline pkg/token/service.go:534]:
```suggestion
        // TODO: Attempting to save token file for Domain %s, Role: %s in %s , ... outPath)
	log.Infof("[New %s Token] Domain: %s, Role: %s", tokenType, domain, role)
	log.Debugf("Saving %s Token[%d bytes] at %s", tokenType, len(rawToken), outPath)
```

File: pkg/token/server.go [hunk 11]
```diff
@@ -78,7 +78,8 @@ func postRoleToken(ts *tokenService, w http.ResponseWriter, r *http.Request) {
 	}
 
 	// create cache key
-	k := CacheKey{Domain: domain, Role: role}
+	// Prioritize checking the cases that are subject to file output.
+	k := CacheKey{Domain: domain, Role: role, WriteFileRequired: true}
```
Comment (mlajkim) [inline pkg/token/server.go:82]: @mlajkim 
Check if this should be true

File: pkg/token/service.go [hunk 12]
```diff
@@ -80,11 +85,15 @@ func New(ctx context.Context, idCfg *config.IdentityConfig) (daemon.Daemon, erro
...
+		// TODO: Rewrite the following if statement as "if tt.isAccessTokenEnabled()..."
 		if tt&mACCESS_TOKEN != 0 {
-			accessTokenCache.Store(CacheKey{Domain: domain, Role: role, MaxExpiry: tokenExpiryInSecond}, &AccessToken{})
+			cacheKey := CacheKey{Domain: domain, Role: role, MaxExpiry: tokenExpiryInSecond, WriteFileRequired: enableWriteFiles}
+			accessTokenCache.Store(cacheKey, &AccessToken{})
 		}
+		// TODO: Rewrite the following if statement as "if tt.isRoleTokenEnabled()..."
 		if tt&mROLE_TOKEN != 0 {
-			roleTokenCache.Store(CacheKey{Domain: domain, Role: role, MinExpiry: tokenExpiryInSecond}, &RoleToken{})
+			cacheKey := CacheKey{Domain: domain, Role: role, MinExpiry: tokenExpiryInSecond, WriteFileRequired: enableWriteFiles}
```
Comment (mlajkim) [inline pkg/token/service.go:95]: Same above

File: pkg/token/service.go [hunk 13]
```diff
@@ -387,108 +391,82 @@ func (ts *tokenService) updateTokenCaches(ctx context.Context, maxElapsedTime ti
...
-		if err := w.AddBytes(outPath, 0644, []byte(rt)); err != nil {
-			return fmt.Errorf("unable to save role token: %w", err)
-		}
-		return nil
-	})
-	if err != nil {
-		return err
 	}
 
+	log.Infof("[Saving %s Token] Domain: %s, Role: %s[%d bytes] in %s", tokenType, domain, role, len(rawToken), outPath)
```
Comment (mlajkim) [inline pkg/token/service.go:466]:
```suggestion
        // TODO: Attempting to ...
	log.Infof("[Saving %s Token] Domain: %s, Role: %s[%d bytes] in %s", tokenType, domain, role, len(rawToken), outPath)
```

File: pkg/token/server.go [hunk 14]
```diff
@@ -96,6 +97,11 @@ func postRoleToken(ts *tokenService, w http.ResponseWriter, r *http.Request) {
 
 	// cache lookup (token TTL must >= 1 minute)
 	rToken := ts.roleTokenCache.Load(k)
+	// The case of not being subject to file output.
+	if rToken == nil {
+		k.WriteFileRequired = false
+		rToken = ts.roleTokenCache.Load(k)
+	}
```
Comment (mlajkim) [inline pkg/token/server.go:104]: How about define a function inside the `TokenCache`
```go
// LoadToDoName loads a token from the given CacheKey.
// It first checks if the token is WriteFileRequired.
// If WriteFileRequired is true, it updates the cache with WriteFileRequired set to true.
// Otherwise, it simply returns the token found in the cache (which can be nil).
func (c *LockedTokenCache) LoadToDoName(k CacheKey) Token {
    // Prioritize WriteFileRequired == true:
    k.WriteFileRequired = true
    token := c.Load(k)
    if token != nil {
        return token
    }

    k.WriteFileRequired = false
    return c.Load(k)
}
```
and do this following instead:
```suggestion
	// The case of not being subject to file output.
	rToken := ts.roleTokenCache.LoadToDoName(k)
```

**Review by mlajkim — COMMENTED** [review]
reiew

File: pkg/token/server.go (hunk 11 above)
Comment (mlajkim) [reply pkg/token/server.go:82]: logically np in the current latest commit.
Yet for better readability for humans, its better to set it `false` and do something else

File: pkg/token/service.go (hunk 10 above)
Comment (y-myajima) [reply pkg/token/service.go]: fixed in https://github.com/AthenZ/k8s-athenz-sia/pull/131/commits/c2df53383b2bc00be4658a1416cae5b696b4697b

File: pkg/token/server.go [hunk 15]
```diff
@@ -255,15 +267,27 @@ func newHandlerFunc(ts *tokenService, timeout time.Duration) http.Handler {
...
 			if ts.tokenType&mACCESS_TOKEN != 0 {
 				aToken = ts.accessTokenCache.Load(k)
+				// The case of not being subject to file output.
+				if aToken == nil {
+					k.WriteFileRequired = false
+					aToken = ts.accessTokenCache.Load(k)
+				}
 				if aToken == nil {
 					errMsg = fmt.Sprintf("domain[%s] role[%s] was not found in cache.", domain, role)
 				}
```
[the remark this answers is not in the document]
Comment (y-myajima) [reply pkg/token/server.go]: fixed in https://github.com/AthenZ/k8s-athenz-sia/pull/131/commits/c2df53383b2bc00be4658a1416cae5b696b4697b

File: pkg/token/service.go [hunk 16]
```diff
@@ -72,6 +72,11 @@ func New(ctx context.Context, idCfg *config.IdentityConfig) (daemon.Daemon, erro
 		log.Info("Skipped token provider initiation")
 		return nil, nil
 	}
+	// TODO: In the next PR, the determination will be made on a per Access Token and Role Token basis.
+	enableWriteFiles := idCfg.TokenDir != ""
+	if !enableWriteFiles {
+		log.Debugf("Skipping to write token files to directory")
```
[the remark this answers is not in the document]
Comment (y-myajima) [reply pkg/token/service.go]: fixed in https://github.com/AthenZ/k8s-athenz-sia/pull/131/commits/c2df53383b2bc00be4658a1416cae5b696b4697b

File: pkg/token/service.go [hunk 17]
```diff
@@ -80,11 +85,15 @@ func New(ctx context.Context, idCfg *config.IdentityConfig) (daemon.Daemon, erro
 	roleTokenCache := NewLockedTokenCache("roletoken", idCfg.Namespace, idCfg.PodName)
 	for _, dr := range idCfg.TargetDomainRoles {
 		domain, role := dr.Domain, dr.Role
+		// TODO: Rewrite the following if statement as "if tt.isAccessTokenEnabled()..."
 		if tt&mACCESS_TOKEN != 0 {
-			accessTokenCache.Store(CacheKey{Domain: domain, Role: role, MaxExpiry: tokenExpiryInSecond}, &AccessToken{})
+			cacheKey := CacheKey{Domain: domain, Role: role, MaxExpiry: tokenExpiryInSecond, WriteFileRequired: enableWriteFiles}
```
[the remark this answers is not in the document]
Comment (y-myajima) [reply pkg/token/service.go]: fixed in https://github.com/AthenZ/k8s-athenz-sia/pull/131/commits/c2df53383b2bc00be4658a1416cae5b696b4697b

File: pkg/token/server.go (hunk 11 above)
Comment (y-myajima) [reply pkg/token/server.go]: fixed in https://github.com/AthenZ/k8s-athenz-sia/pull/131/commits/c2df53383b2bc00be4658a1416cae5b696b4697b

File: pkg/token/service.go (hunk 12 above)
Comment (y-myajima) [reply pkg/token/service.go]: fixed in https://github.com/AthenZ/k8s-athenz-sia/pull/131/commits/c2df53383b2bc00be4658a1416cae5b696b4697b

File: pkg/token/server.go (hunk 14 above)
Comment (y-myajima) [reply pkg/token/server.go]: fixed in https://github.com/AthenZ/k8s-athenz-sia/pull/131/commits/c2df53383b2bc00be4658a1416cae5b696b4697b

File: pkg/token/server.go [hunk 18]
```diff
@@ -78,7 +78,7 @@ func postRoleToken(ts *tokenService, w http.ResponseWriter, r *http.Request) {
 	}
 
 	// create cache key
-	k := CacheKey{Domain: domain, Role: role}
+	k := CacheKey{Domain: domain, Role: role, WriteFileRequired: false}
```
Comment (mlajkim) [inline pkg/token/server.go:81]: no need to declare at this point:
```suggestion
	k := CacheKey{Domain: domain, Role: role}
```

File: pkg/token/server.go [hunk 19]
```diff
@@ -165,7 +166,7 @@ func postAccessToken(ts *tokenService, w http.ResponseWriter, r *http.Request) {
...
```
Comment (mlajkim) [inline pkg/token/server.go:169]: no need to declare at this point:
```suggestion
	k := CacheKey{Domain: domain, Role: role}
```

**Review by mlajkim — COMMENTED** [review]
review

File: pkg/token/server.go (hunk 19 above)
Comment (y-myajima) [reply pkg/token/server.go]: fixed in: https://github.com/AthenZ/k8s-athenz-sia/pull/131/commits/65185656cd52eeb18ad050fd0a5a25a5033b61e0

File: pkg/token/server.go (hunk 18 above)
Comment (y-myajima) [reply pkg/token/server.go]: fixed in: https://github.com/AthenZ/k8s-athenz-sia/pull/131/commits/65185656cd52eeb18ad050fd0a5a25a5033b61e0

File: pkg/token/cache.go [hunk 20]
```diff
@@ -108,6 +110,29 @@ func (c *LockedTokenCache) Store(k CacheKey, t Token) {
...
+	}
+	key.WriteFileRequired = false
+	t = c.Load(key)
+	if t != nil {
+		return key, t
+	}
+	// If there is no cache hit, it returns the cache key specified in the arguments as is.
+	return k, nil
+}
+
```
[the remark this answers is not in the document]
Comment (y-myajima) [reply pkg/token/cache.go:141]: fixed in: https://github.com/AthenZ/k8s-athenz-sia/pull/131/commits/65185656cd52eeb18ad050fd0a5a25a5033b61e0

File: pkg/token/service.go [hunk 21]
```diff
@@ -126,16 +133,15 @@ func New(ctx context.Context, idCfg *config.IdentityConfig) (daemon.Daemon, erro
...
-		errs := ts.updateTokenCaches(ctx, config.DEFAULT_MAX_ELAPSED_TIME_ON_INIT)
+	// write tokens as files only if it is non-init mode OR TOKEN_DIR is set
+	// If it is in refresh mode, when requesting tokens using the REST API for the domains and roles specified in TARGET_DOMAIN_ROLES,
+	// the cache is updated to ensure a cache hit from the first request.
+	if !idCfg.Init || enableWriteFiles {
+		errs := ts.updateTokenCachesAndWriteFiles(ctx, config.DEFAULT_MAX_ELAPSED_TIME_ON_INIT)
 		// TODO: if cap(errs) == len(errs), implies all token updates failed, should be fatal
 		for _, err := range errs {
 			log.Errorf("Failed to refresh tokens after multiple retries: %s", err.Error())
 		}
```
[the remark this answers is not in the document]
Comment (y-myajima) [reply pkg/token/service.go]: fixed in: https://github.com/AthenZ/k8s-athenz-sia/pull/131/commits/65185656cd52eeb18ad050fd0a5a25a5033b61e0

File: pkg/token/cache.go [hunk 22]
```diff
@@ -28,6 +28,7 @@ import (
 
 type TokenCache interface {
 	Store(k CacheKey, t Token)
+	Search(k CacheKey) (CacheKey, Token)
 	Load(k CacheKey) Token
```
Comment (mlajkim) [inline pkg/token/cache.go:32]: mb:
```suggestion
         Load(k CacheKey) Token
	Search(k CacheKey) (CacheKey, Token)
```

**Review by mlajkim — APPROVED** [review]
ty for the amazing work
